content/blog/2019-12-18-novm/index.md

```diff
@@ -0,0 +1,317 @@
+and warm up times for each program.
+
+<img src="stackcheckparsec.png" style="width:100%"/>
+
+With these four benchmarks, there was almost no impact of applying the split-stack option.
+We should note that the StreamCluster benchmark actually _sped up_ when using split-stack;
+likely this is some sort of memory alignment effect à la [Mytcowicz et al.](https://dl.acm.org/citation.cfm?id=1508275).
+In any case, we should probably consider this impact to be negligible.
+
+Of these benchmarks, only Ferret actually required dynamically allocating stack space.
+Each of these allocations was 618 KB, which is a potential concern. It is unclear, in a real
+system using only physical addressing, whether or not allocations of this size would
+be frequently servicable or not. I hypothesize that real systems with many gigabytes or terabytes
+of memory with even severe fragmentation will be able to regularly respond to allocations
+in the kilobyte range; however, proving this is future work.
+
+# Large Object Allocations
+
+The other major modification to programs would be supporting large memory allocations.
+Since it is probably unreliable to request very large contiguous memory regions, we must adopt a new
+strategy. To evaluate the potential impact of these changes, we modified the Blackscholes
+benchmark from the PARSEC suite. Blackscholes uses two dynamically allocated arrays,
+which we replaced with custom `Array` objects that we implemented to use only fixed size
+allocations. We chose to modify this application not only because it consists of a single,
+easily-modifiable source file, but also because it iterates through large arrays
+and is very likely to be negatively impacted by array access latency and spatial locality within the array.
+
+## Custom Array Implementation
+
+We implemented our `Array` using a tree-based datastructure that mimics
+the functionality of page tables in a virtual address environment.
+We support up to three levels of allocation, where the final
+level contains data and all previous levels contain pointers to
+other pages. In the following diagram, the **L1** page contains
+pointers to **L2** pages, which contain pointers to the **L3** pages.
+Each **L3** page contains the actual object array data.
+
+<img src="treedata.png" style="width:100%"/>
+
+As an optimization, the constructor for `Array` determines how many
+levels of pages are required to store all of the data. For instance,
+in the event the data fits in a single page, then the **L1** page will hold data
+and no **L2** or **L3** pages will be allocated.
+
+## Evaluating Tree Overheads
+
+We had to modify the Blackscholes benchmark slightly to replace the calls
+to `malloc` to use our C++ `Array` objects. This only involved modifying
+the array allocation and deallocation since we overloaded the `[]` operator
+for normal array dereferences.
+Blackscholes only uses pointer arithmetic for allocation purposes so we didn't
+need to modify any other instructions.
```

Comment:
Overall, this exercise seems to imply that this idea is probably best suited to languages with nontrivial runtimes—like Java or .NET or whatever, where you can imagine transparently swapping in different implementations of the basic standard-library data structures that everybody uses.